### PR TITLE
Code Refactoring

### PR DESCRIPTION
**Pull request by khaledosman**
- replace confusing Array.map implementation that mutates the state object with Array.reduce
- use memo & PureComponent
- use rest parameters instead of "a,b,c,d,e,f" params and fix using ++ on constant variable

### CODE REVIEW

File: createModalStack.js [hunk 1]
```diff
@@ -23,18 +23,16 @@ export default function(config: Config, customDefaultOptions: Options): Stack {
     total: 0,
   }
 
-  Object.entries(config).map((entry, index) => {
+  return Object.entries(config).reduce((accum, entry, index) => {
```
Comment (CharlesMangwa) [inline createModalStack.js]: nice one ✨👌

**Comment by lucasmotta** [comment]
Some pretty solid changes on this PR - @khaledosman do you think you could rebase your changes and resolve the conflicts? Then @CharlesMangwa could probably help approving and merging!

**Comment by lucasmotta** [comment]
Hey @CharlesMangwa, any chances that this could get merged and maybe a new version release? There are some methods like `closeAllModals` that is only available on master at  the moment.

**Comment by CharlesMangwa** [comment]
Hi @lucasmotta! First of all: thanks for checking out `react-native-modalfy`! Almost all the changes @khaledosman brought are already available in the current `0.1.1`. I didn't have time to work on RNM and they're still some changes that actually break the library, that's why this PR hasn't been merged yet.

About the changes and new methods you're seeing on master: they're breaking changes so that imply that the next version will be a major one (`1.0.0`). We'd like to bring unit tests, update the docs, etc before getting there which might take a tiny bit because we've some other priorities right now. However, if you really need those new methods, feel free to install the library from the latest version on master and I could guide you through the new things if needed!
We'll do our best to get `1.0.0` to you in the as soon as practicable

**Comment by CharlesMangwa** [comment]
@lucasmotta Just a heads up for you: `[react-native-modalfy@1.0.0-beta.0](https://www.npmjs.com/package/react-native-modalfy/v/1.0.0-beta.0)` is out with all the new features! The documentation will be updated as soon as we can but feel free to browse the source code or ask us for any help in the meantime!

**Pull request closed without merging**